docs/source/notes/rref.rst

```diff
@@ -0,0 +1,304 @@
+Remote Reference Protocol
+=========================
+
+.. warning::
+  The :ref:`rref` API is experimental and subject to change.
+
+This note describes the design details of Remote Reference protocol and walks
+through message flows in different scenarios. Make sure you're familiar with the
+:ref:`distributed-rpc-framework` before proceeding.
+
+Background
+^^^^^^^^^^
+
+RRef stands for Remote REFerence. It is a reference of an object which locates
+on the local or a remote worker, and transparently handles reference counting
+under the hood. Conceptually, it can be considered as a distributed shared
+pointer. Applications can create an RRef by calling
+`torch.distributed.rpc.remote`. Each RRef is owned by the callee worker of the
+`torch.distributed.rpc.remote` call (i.e., owner) and can be used by multiple
+users. The owner stores the real data and keeps track of the global reference
+count. Every RRef can be uniquely identified by a global ``RRefId``,
+which is assigned at the time of creation on the caller of the
+`torch.distributed.rpc.remote` call.
+
+On the owner worker, there is only one ``OwnerRRef`` instance, which contains
+the real data, while on user workers, there can be as many ``UserRRefs`` as
+necessary, and ``UserRRef`` does not hold the data. All usage on the owner will
+retrieve the unique ``OwnerRRef`` instance using the globally unique ``RRefId``.
+A ``UserRRef`` will be created when it is used as an argument or return value in
+``rpc`` or ``remote`` invocation, and the owner will be notified according to
+update the reference count. An ``OwnerRRef`` and its data will be deleted when
```

Comment:
Nit, are these classes clickable links in the docs?

Reply:
No, they are not currently, let me make changes, thanks!